### PR TITLE
Ruby: fix some synchronization code in server implementation

### PR DESCRIPTION
**Pull request by murgatroid99**
I think this will fix the timeouts we've been seeing on Jenkins.

### CODE REVIEW

File: src/ruby/lib/grpc/generic/rpc_server.rb [hunk 1]
```diff
@@ -467,11 +476,15 @@ def new_active_server_call(an_rpc)
...
-      @rpc_descs ||= {}
+      @handle_mutex.synchronize do
+        return @rpc_descs ||= {}
+      end
     end
 
     def rpc_handlers
-      @rpc_handlers ||= {}
+      @handle_mutex.synchronize do
+        @rpc_handlers ||= {}
```
Comment (nicolasnoble) [inline src/ruby/lib/grpc/generic/rpc_server.rb]: Why the discrepancy between that and line 480 ? (with and without return)

**Comment by nicolasnoble** [comment]
This is LGTM, but Jenkins doesn't want to make that green because reasons.

File: src/ruby/lib/grpc/generic/rpc_server.rb [hunk 2]
```diff
@@ -275,27 +277,42 @@ def initialize(pool_size:DEFAULT_POOL_SIZE,
...
+        @running_state = target_state
+      else
+        fail "Bad server state transition: #{@running_state}->#{target_state}"
       end
     end
 
-    # determines if the server is currently running
     def running?
-      @running
+      running_state == :running
```
Comment (jtattermusch) [inline src/ruby/lib/grpc/generic/rpc_server.rb]: shouldn't we synchronize the reads as well?

Comment (nicolasnoble) [inline src/ruby/lib/grpc/generic/rpc_server.rb]: Actually, these one should be synchronized, yes, you are right.

File: src/ruby/lib/grpc/generic/rpc_server.rb [hunk 3]
```diff
@@ -413,9 +428,9 @@ def implemented?(an_rpc)
 
     # handles calls to the server
     def loop_handle_server_calls
-      fail 'not running' unless @running
+      fail 'not started' if running_state == :not_started
       loop_tag = Object.new
-      until stopped?
+      while running_state == :running
```
Comment (jtattermusch) [inline src/ruby/lib/grpc/generic/rpc_server.rb]: is there a concept of volatile in ruby? technically just reading  running_state is unsafe.

Comment (murgatroid99) [inline src/ruby/lib/grpc/generic/rpc_server.rb]: This is the method running_state, which is synchronized.

File: src/ruby/lib/grpc/generic/rpc_server.rb (hunk 2 above)
Comment (nicolasnoble) [inline src/ruby/lib/grpc/generic/rpc_server.rb]: 'doh, you're right... I always get confused :P

File: src/ruby/lib/grpc/generic/rpc_server.rb (hunk 3 above)
Comment (jtattermusch) [inline src/ruby/lib/grpc/generic/rpc_server.rb]: Ah, thanks for explaining.

**Comment by nicolasnoble** [comment]
Merging that one then.

**Pull request merged**